src/v/cluster/members_backend.cc

```diff
@@ -103,79 +128,257 @@ void members_backend::start_reconciliation_loop() {
     });
 }
 
+bool is_in_replica_set(
+  const std::vector<model::broker_shard>& replicas, model::node_id id) {
+    return std::any_of(
+      replicas.cbegin(), replicas.cend(), [id](const model::broker_shard& bs) {
+          return id == bs.node_id;
+      });
+}
+
+void members_backend::calculate_reallocations(update_meta& meta) {
+    switch (meta.update.type) {
+    case members_manager::node_update_type::decommissioned:
+        // reallocate all partitons for which any of replicas is placed on
+        // decommissioned node
+        for (const auto& [tp_ns, cfg] : _topics.local().topics_map()) {
+            for (auto& pas : cfg.assignments) {
+                if (is_in_replica_set(pas.replicas, meta.update.id)) {
+                    partition_reallocation reallocation(
+                      model::ntp(tp_ns.ns, tp_ns.tp, pas.id),
+                      pas.replicas.size());
+                    reallocation.replicas_to_remove.emplace(meta.update.id);
+                    meta.partition_reallocations.push_back(
+                      std::move(reallocation));
+                }
+            }
+        }
+        return;
+    case members_manager::node_update_type::added:
+        calculate_reallocations_after_node_added(meta);
+        return;
+    default:
+        return;
+    }
+}
+/**
+ * Simple helper class represeting how many replicas have to be moved from the
+ * node.
+ */
+struct replicas_to_move {
+    replicas_to_move(model::node_id id, uint32_t left)
+      : id(id)
+      , left_to_move(left) {}
+
+    model::node_id id;
+    uint32_t left_to_move;
+    friend std::ostream& operator<<(std::ostream& o, const replicas_to_move& r);
+};
+
+std::ostream& operator<<(std::ostream& o, const replicas_to_move& r) {
+    fmt::print(o, "{{id: {}, left_to_move: {}}}", r.id, r.left_to_move);
+    return o;
+};
+
+void members_backend::calculate_reallocations_after_node_added(
+  members_backend::update_meta& meta) {
+    auto& topics = _topics.local().topics_map();
+    struct node_info {
+        size_t replicas_count;
+        size_t max_capacity;
+    };
+    // 1. count current node allocations
+    absl::node_hash_map<model::node_id, node_info> node_replicas;
+
+    size_t total_replicas = 0;
+    for (const auto& [id, n] : _allocator.local().state().allocation_nodes()) {
+        if (!node_replicas.contains(id)) {
+            node_replicas.emplace(
+              id,
+              node_info{
+                .replicas_count = 0,
+                .max_capacity = n->max_capacity(),
+              });
+        }
+        auto it = node_replicas.find(id);
+        it->second.replicas_count += n->allocated_partitions();
+        total_replicas += n->allocated_partitions();
+    }
+
+    // 2. calculate number of replicas per node leading to even replica per node
+    // distribution
```

Comment:
Shouldn't we take into account the size of the node (number of shards) and sizes of individual partitions? Even replica per node distribution might not be optimal. Or this is a first iteration and it'll get more elaborate eventually?

Reply:
yes I think we'll be adding more dimensions to the strategy over time